enf-reducer/n-ary-gather-junctors.metta

```diff
@@ -0,0 +1,72 @@
+ (= (concat-tuple $xs $ys)
+    (if (== $xs ()) $ys
+        (let* (($head (car-atom $xs))  
+               ($tail (cdr-atom $xs)) 
+               ($tail-new (concat-tuple $tail $ys))  
+              )
+          (cons-atom $head $tail-new)  
+        )
+    )
+)
+
+(= (fold $f $tuple $acc)
+    (if (== $tuple ()) $acc
+        (fold $f (cdr-atom $tuple) ($f $acc (car-atom $tuple)))
+    )
+)
+; (: gather-junctors (-> Expression Expression))
+(= (gather-junctors $expr $prev)
+    (
+        if (== (get-metatype $expr) Expression)
+                (
+            let* (
+                ($op (car-atom $expr)) 
+                ($tuple (cdr-atom $expr))
+                ($newTuple (fold concat-tuple (collapse (gather-junctors (superpose $tuple) $op)) () ))
+                ($tupleWithOp (cons-atom $op $newTuple))
+            )
+            (
+                if (== $op $prev)
+                    $newTuple
+                    (
+                        if (== $prev ())
+                            $tupleWithOp
+                            ($tupleWithOp)
+                    )
+
+            )
+        )
+        
+        ($expr)
+    )
+)       
+
+
+
+
+(= (map $f $expr $p)
+   (if (== $expr ()) 
+       ()  
+       (let* (($head (car-atom $expr))
+              ($tail (cdr-atom $expr))
+              ($params (cons-atom $head $p))
+              ($headWithParams (cons-atom $f $params)) 
+              ($tailWithParams (map $f $tail $p))
+             
+             )
+         (cons-atom $headWithParams $tailWithParams)  
+       )
+   )
+)
+
+ 
+(= (gatherJunctors $expr) (gather-junctors $expr ()))
```

Comment:
What is the purpose of this line here?

Reply:
the function gather-junctors function takes two arguments -- the expression itself and an accumulator . but when the function is called for the first time we have to pass an empty tuple for the accumulator . I thought this would make the code unreadable when this function is used .

the function below is just a wrapper that can be used to avoid unnecessary passing of empty arguments